### PR TITLE
Fix #2639: Refuse to import modules without module support

### LINKED ISSUE #2639: The linker should refuse to build if @JSImport is used with NoModule
I tried to import electron in my application. But build is failing to import.

Full problem is explained here : 

```

import scala.scalajs.js
import js.annotation._
import scala.scalajs.js.annotation.JSExport

@JSExport("CloudConnectionComponent.App")
object CloudConnectionComponent extends js.JSApp {

  @ScalaJSDefined
  trait BrowserWindowOptions extends js.Object {
    def width: Double
    def height: Double
  }

  @js.native
  @JSImport("electron", "BrowserWindow")
  class BrowserWindow(options: BrowserWindowOptions) extends js.Object {
    def on(eventName: String, handler: js.Function0[Any]): Unit = js.native
    def openDevTools():Unit = js.native
    def loadUrl(url:String) :Unit = js.native
  }

  val dropboxOAuthUrl = "https://www.dropbox.com/1/oauth2/authorize?client_id=8343b03llcys1pw&response_type=token&redirect_uri=http://localhost:3000/auth/db/callback&state=433"
  var mainWindow:BrowserWindow = null

  @JSExport
  def main() = {
    mainWindow = new BrowserWindow(new BrowserWindowOptions {
      val width = 800.0
      val height = 600.0
    })

    mainWindow.loadUrl(dropboxOAuthUrl)
    mainWindow.openDevTools()
    mainWindow.on("closed", () => {
      mainWindow = null
    })
  }

}
```

Error : 

```
target/scala-2.11/cloud-manager-app.js.map -> src/main/electron/cloud-manager-app.js.map
App threw an error during load
ReferenceError: $i_electron is not defined
    at ta (/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/src/main/electron/file:/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/src/main/scala/CloudConnectionComponent.scala:31:18
    )
    at T.main (/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/src/main/electron/file:/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/src/main/scala/CloudConnectionComponent.scala:29:4)
    at Object.<anonymous> (/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/src/main/electron/cloud-manager-app-launcher.js:22:50)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at loadApplicationPackage (/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/default_app.asar/main.js:288:12)
    at Object.<anonymous> (/Users/rajeevprasanna/Desktop/cloudManager/CloudManagerApp/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/default_app.asar/main.js:330:5)

```

[http://stackoverflow.com/questions/40224123/import-electron-browser-window-in-scala-js/40225351](http://stackoverflow.com/questions/40224123/import-electron-browser-window-in-scala-js/40225351)

Project code with example is available here : [https://github.com/rajeevprasanna/CloudManagerApp](https://github.com/rajeevprasanna/CloudManagerApp)

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Review by @sjrd

**Comment by gzm0** [comment]
I manually checked that this triggers if a `@JSImport` is used.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala [hunk 1]
```diff
@@ -116,9 +117,29 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
...
+          val displayName = decodeClassName(classDef.encodedName)
+          (displayName, module)
+        }
+
+        for {
+          (className, module) <- imports
+        } {
+          logger.error(s"$className needs to be imported from module " +
+              s"'$module' but module support is disabled.")
+        }
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala]: The `Emitter`, especially the non-cached parts, is quite sensitive to the number of allocations (discovered by @nicolasstucki some time ago in a profiling & optimization session). We should make a better effort not to introduce all those tuple allocations and pattern-matching-based `withFilter`s. Consider something like this:
```scala
var importsFound: Boolean = false

for (classDef <- orderedClasses) {
  classDef.jsNativeSpec match {
    case Some(JSNativeLoadSpec.Import(module, _)) =>
      val displayName = decodeClassName(classDef.encodedName)
      logger.error(s"$className needs to be imported from module " +
          s"'$module' but module support is disabled.")
      importsFound = true
    case _ =>
      // ok
}

if (importsFound) {
  sys.error(...)
}
```

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala [hunk 2]
```diff
@@ -116,9 +117,29 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
...
+
+        if (imports.nonEmpty) {
+          sys.error("There were module imports, but module support is " +
+            "disabled.\nTo enable module support, set scalaJSModuleKind := " +
+            "ModuleKind.CommonJSModule.")
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala]: 4 spaces.

**Comment by gzm0** [comment]
Updated.

**Pull request merged**